phases/11-llm-engineering/16-langgraph-state-machines/code/main.py

```
@@ -0,0 +1,169 @@
+# Tools ----------------------------------------------------------------------
+
+
+@tool
+def calculator(expression: str) -> str:
+    """Evaluate a Python arithmetic expression like '2 + 2 * 3'. Returns the
+    result as a string."""
+    allowed = set("0123456789+-*/(). ")
+    if not set(expression) <= allowed:
+        return "ERROR: only digits and + - * / ( ) are allowed"
+    try:
+        return str(eval(expression, {"__builtins__": {}}, {}))
```

Ruff (0.15.10)

[error] Use of possibly insecure function; consider using `ast.literal_eval`

(S307)